### ISSUE TITLE
add the BIOM file formats to skbio.io

### ISSUE THREAD

**Issue by jairideout** [biocore/scikit-bio#828]
Depends on:

- [ ] I/O aliases (#634)

**Comment by wasade** [comment]
Updating following our email conversation:

- [ ] `read(a_biom_file, into=biom.Table)` should produce a `biom.Table` object
- [ ] `write(a_table, format='biom-format-1.0.0', into=a_ascii_file)` should write a BIOM format v1.0.0 table (e.g., JSON)
- [ ] `write(a_table, format='biom-format-2.1.0', into=a_hdf5_file)` should write a BIOM format v2.1.0 table (e.g., HDF5) 
- [ ] `Table.read` and `Table.write` should work as expected

cc @gregcaporaso @ebolyen

**Comment by ebolyen** [comment]
Include:
 - [ ] `read(file, into=xray.Dataset)`
 - [ ] `write(xray.Dataset, format='biom-format-2.1.0')`

**Comment by ElDeveloper** [comment]
@ebolyen and @wasade, I'm sure you guys may have discussed at length but I wanted to check. With `xray.Dataset`s, how is metadata going to be handled?. All in all these seem like great additions!

**Comment by ebolyen** [comment]
@ElDeveloper We haven't actually discussed that in any detail. I think `xray` gives an `attrs` property for metadata on a dimension, but I'm not sure how we would choose to use that specifically.

**Comment by wasade** [comment]
Theres very little (ie none) on the metadata of an in memory biom table at
this point. That should probably be changed going forward as we don't find
out about metadata issues until write. So, ya, we haven't but should.
Another possibility are pandas tables per axis?

**Comment by ebolyen** [comment]
Long-term that might a good solution, but it would force a composite object. Which is probably desirable to avoid until we know more about how this structure behaves and have had a chance to interact with it in non-trivial ways.

**Comment by wasade** [comment]
biom-format v1.0.0 is ascii, while biom-format v2.1.0 is binary. `create_format` supports the encoding type, but I don't believe it supports both binary and ascii at the same time. One strategy would be to have two formats, but wasn't sure if that was the best way to proceed here.

**Comment by ebolyen** [comment]
@wasade We would need two formats anyhow as the sniffers should be independent. Then resolving the version is actually the same as figuring out what format it is in the first place. We don't have any real precedent for naming the format with a version, so anything sensible should work.

Also I'm not even sure how the registry *could* support both kinds of files for the same format, they are different objects entirely in Python.

**Comment by wasade** [comment]
I guess I'm not sure how to resolve the ambiguity of `table.write(fh)`, is that biom-format-2.1.0, biom-format-1.0.0, or is dependent on the type of `fh`? Reading in I think is fine by just having multiple formats, but the writing is where I'm still unsure on as `Table` has two different ascii representations and a binary one.

**Comment by ebolyen** [comment]
I think I might be misunderstanding the problem.

Those formats would all be bound to the format name. So someone would say:
``skbio.io.write(table, into=fh, format='biom-format-2.1.0')`` or whichever one they wanted.

**Comment by wasade** [comment]
Ah, okay. So it would `raise` if `fh` was incompatible with the requested format?

**Comment by ebolyen** [comment]
yes, the registry enforces the constraints you place on it. So it won't work if you give it an incompatible file source.
